src/main/kotlin/org/wycliffeassociates/otter/jvm/persistence/repositories/ResourceContainerRepository.kt

```diff
@@ -0,0 +1,154 @@
+            database.transaction { dsl ->
+                val language = LanguageMapper().mapFromEntity(languageDao.fetchBySlug(languageSlug, dsl))
+                val metadata = rc.manifest.dublinCore.mapToMetadata(rc.dir, language)
+                val dublinCoreFk = resourceMetadataDao.insert(ResourceMetadataMapper().mapToEntity(metadata), dsl)
+
+                // TODO: Should this only happen if we are importing a "help" RC? What if the help is imported first,
+                // TODO: ... then another related bundle RC is imported?
+                val relatedDublinCoreIds: List<Int> =
+                        linkRelatedResourceContainers(dublinCoreFk, rc.manifest.dublinCore.relation, dsl)
+
+                // TODO: Probably shouldn't hardcode "help"
+                if (rc.type() == "help") {
+                    relatedDublinCoreIds.forEach { relatedId ->
+                        val ih = ImportHelper(dublinCoreFk, relatedId, dsl)
+                        ih.importCollection(null, rcTree)
+                    }
+                } else {
+                    val ih = ImportHelper(dublinCoreFk, null, dsl)
+                    ih.importCollection(null, rcTree)
+                }
+            }
+        }.subscribeOn(Schedulers.io())
+    }
+
+    private fun linkRelatedResourceContainers(
+            dublinCoreFk: Int,
+            relations: List<String>,
+            dsl: DSLContext)
+            : List<Int> {
+        val relatedIds = mutableListOf<Int>()
+        relations.forEach { relation ->
+            val parts = relation.split('/')
+            resourceMetadataDao.fetchByLanguageAndIdentifier(parts[0], parts[1], dsl)
+                    ?.let { relatedDublinCore ->
+                        // TODO: Only add link if it doesn't exist already
+                        resourceMetadataDao.addLink(dublinCoreFk, relatedDublinCore.id, dsl)
+                        relatedIds.add(relatedDublinCore.id)
+                    }
+        }
+        return relatedIds
+    }
+
+    inner class ImportHelper(val dublinCoreId: Int, val relatedBundleDublinCoreId: Int?, val dsl: DSLContext, var collectionsImported: Int = 0) {
+
+        fun importCollection(parentId: Int?, node: Tree) {
+            (node.value as? Collection)?.let { collection ->
+                when (relatedBundleDublinCoreId) {
+                    null -> addCollection(collection, parentId)
+                    else -> findCollectionId(collection, relatedBundleDublinCoreId)
+                }
+            }.let { collectionId ->
+                // TODO: If we don't find a corresponding collection, we continue on, passing null to collectionId.
+                // TODO ... Eventually, contents will not be created if there is no parentId. This will happen for front
+                // TODO ... matter until we have another solution.
+                for (childNode in node.children) {
+                    importNode(collectionId, childNode)
```

Comment:
I can't think of a situation where the types don't match (like, the children are either all content, or all collections), so I suppose one thing we could do is rather than iterate and import nodes one at a time, you could find the type here, and insert all into a list, and then insert the list into the database at once.

If the assumption doesn't hold (and they could be mixed types), then you could filter them out into two lists.

I think we'd probably have other issues if they did mix and match anyway though?

Reply:
It's also possible this wouldn't affect anything anyway because jooq might hold off on inserting?

Reply:
no actually, it probably can't wait, and the insert is returning an id right away

I forgot- trying to optimize this stuff is probably going to mean doing things down inside the dao level, not at a high level like this

Reply:
That might work- perhaps we can create a story to accomplish batch inserting later? (By later I mean soon)